zkevm-circuits/src/evm_circuit/execution/mul_div_mod_shl_shr.rs

```diff
@@ -0,0 +1,351 @@
+        let [pop1, pop2, push] = indices.map(|idx| block.rws[idx].stack_value());
+
+        // Shift values are only used for opcode SHL and SHR.
+        let shf0 = pop1.to_le_bytes()[0];
+        let shf_lt256 = pop1
+            .to_le_bytes()
+            .iter()
+            .fold(0, |acc, val| acc + *val as u128)
+            - shf0 as u128;
+
+        let (quotient, divisor, remainder, dividend) = match step.opcode.unwrap() {
+            OpcodeId::MUL => (pop1, pop2, U256::from(0), push),
+            OpcodeId::DIV => (push, pop2, pop1 - push * pop2, pop1),
+            OpcodeId::MOD => (
+                if pop2.is_zero() {
+                    U256::from(0)
+                } else {
+                    pop1 / pop2
+                },
+                pop2,
+                if pop2.is_zero() { pop1 } else { push },
+                pop1,
+            ),
+            OpcodeId::SHL => (
+                pop2,
+                if shf_lt256 == 0 {
+                    U256::from(1) << shf0
+                } else {
+                    U256::from(0)
+                },
+                U256::from(0),
+                push,
+            ),
+            OpcodeId::SHR => {
+                let divisor = if shf_lt256 == 0 {
+                    U256::from(1) << shf0
+                } else {
+                    U256::from(0)
+                };
+                (push, divisor, pop2 - push * divisor, pop2)
+            }
+            _ => unreachable!(),
+        };
+        self.words[0].assign(region, offset, Some(quotient.to_le_bytes()))?;
+        self.words[1].assign(region, offset, Some(divisor.to_le_bytes()))?;
+        self.words[2].assign(region, offset, Some(remainder.to_le_bytes()))?;
+        self.words[3].assign(region, offset, Some(dividend.to_le_bytes()))?;
+        self.words[4].assign(region, offset, Some(pop1.to_le_bytes()))?;
+        self.mul_add_words
+            .assign(region, offset, [quotient, divisor, remainder, dividend])?;
+        let divisor_sum = (0..32).fold(0, |acc, idx| acc + divisor.byte(idx) as u64);
+        self.divisor_is_zero
+            .assign(region, offset, F::from(divisor_sum))?;
+        let remainder_sum = (0..32).fold(0, |acc, idx| acc + remainder.byte(idx) as u64);
+        self.remainder_is_zero
+            .assign(region, offset, F::from(remainder_sum))?;
+        self.remainder_lt_divisor
+            .assign(region, offset, remainder, divisor)
+    }
+}
+
+// The opcode value for MUL, DIV, MOD, SHL and SHR are 2, 4, 6, 0x1b and 0x1c.
+// When the opcode is MUL, the result of below formula is 5200:
+// (DIV - opcode) * (MOD- opcode) * (SHL - opcode) * (SHR - opcode)
+// To make `is_mul` be either 0 or 1, the result needs to be divided by 5200,
+// which is equivalent to multiply it by inversion of 5200.
+// And calculate `is_div`, `is_mod`, `is_shl` and `is_shr` respectively.
+#[inline]
+fn is_mul<F: Field>(opcode: &Cell<F>) -> Expression<F> {
```

Comment:
You can use `util::math_gadget::generate_lagrange_base_polynomial` for all of these.